### ISSUE TITLE
Lab2 review

### ISSUE THREAD

**Issue by GiorgioBongiovanni** [nrbeo/CI2024_lab2#1]
The TSP problem focuses on finding the shortest Hamiltonian cycle in a complete graph, with the following fundamental characteristics:
- Cycle structure: The route must be cyclic, meaning the first and last nodes must coincide.
- Importance of adjacency: The quality of a solution depends on the proximity of cities in the route.
- Permutation constraint: Each city must be visited exactly once.

From my point of view, these constraints make some genetic operators more suitable than others. For example:
Inver-over crossover is designed to preserve favorable adjacencies and inversion mutation modifies the route in a localized manner maintaining the proximity of cities.
The developed genetic algorithm aims to find a more accurate solution to the TSP problem compared to the greedy algorithm. However, the best path found, for example, in the instance concerning Italy (6933.95 km) is significantly far from the known optimal solution (4172.76 km). This difference suggests potential improvements in the implemented approach. Therefore, while this genetic algorithm provides a solid foundation for solving the TSP, it requires significant enhancements to approach the optimal solution.

The current approach uses:
- Order Crossover (OX): While suitable for permutation problems, it does not leverage semantic information related to adjacency. In TSP, the connections between cities are more critical than their relative positions.
- Multiple mutations (swap, scramble): Although flexible, these mutations may introduce excessive variation (scramble) or insufficient exploration (swap), reducing the efficiency of exploration.

Instead, the inver-over crossover combines inversion and crossover leveraging the structure of the TSP:
- Avoiding adjacency disruption: Inver-over preserves favorable connections between cities in two ways:
It uses information from a second individual to guide the choice of the subsequence to invert.
It selects cities based on their proximity in the graph.
- Targeted exploration: It generates variations of the route by inverting only a localized portion of the genome. This allows more "intelligent" exploration compared to random mutations.
By using inver-over, the algorithm should better preserve local connections between cities while maintaining the cyclic nature of TSP solutions.

Here, I write an example pseudocode for integrating it into the algorithm:
```python
def inver_over_crossover(parent1, population, p):
    offspring = parent1[:]
    current_city = random.choice(offspring)  # Select a random city
    while True:
        # Decide whether to use another city from the same parent or another individual
        if random.random() < p:
            next_city = random.choice([c for c in offspring if c != current_city])
        else:
            random_individual = random.choice(population)
            next_city = random_individual[(random_individual.index(current_city) + 1) % len(random_individual)]
        
        # Terminate if current_city and next_city are already adjacent
        if abs(offspring.index(current_city) - offspring.index(next_city)) == 1:
            break
        
        # Reverse the subsequence
        i, j = offspring.index(current_city), offspring.index(next_city)
        if i > j:
            i, j = j, i
        offspring[i+1:j+1] = offspring[i+1:j+1][::-1]
        
        # Update the current city
        current_city = next_city
    return offspring
```

In addition, inversion mutation is a localized mutation that changes the order of cities between two points while maintaining the global structure of the route. It is particularly useful in TSP for the following reasons:
- Preserving geographical proximity: Reversing a subsequence keeps local connections between cities reducing the risk of creating suboptimal solutions.
- Balanced exploration and optimization: Unlike swap or scramble, inversion introduces significant changes without drastically altering the solution.

```python
def inversion_mutation(route):
    mutated_route = route[:]
    i, j = sorted(random.sample(range(1, len(mutated_route) - 1), 2))  # Avoid modifying the start/end city
    mutated_route[i:j+1] = reversed(mutated_route[i:j+1])
    return mutated_route
```

Other suggestions for improvement could be:
Retaining the top 10% of the best individuals (elitism) may reduce diversity. 
A more balanced approach could combine elitism with random replacements to reintroduce variability.
Secondly, adding a 2-opt phase as a local search operator can significantly improve the quality of solutions.
Each individual can be refined locally before its fitness evaluation.
Thirdly, making the probabilities of crossover and mutation adaptive—for instance, based on the quality of the solutions found—can improve performance (more exploration in the early stages, more exploitation in the later stages).

In conclusion, the use of the Inver-Over crossover and the inversion mutation would better align the algorithm with the semantics of the TSP. 
Although the current implementation provides a solid foundation, adopting these specific operators and achieving a better balance between exploration and exploitation could bring the solutions closer to the optimal ones.
